client/java/src/main/java/io/openlineage/client/circuitBreaker/JavaRuntimeCircuitBreaker.java

```diff
@@ -0,0 +1,153 @@
+      new ThreadLocal<Long>() {
+        @Override
+        protected Long initialValue() {
+          return getGCCpuTimeNS();
+        }
+      };
+  private final ThreadLocal<Long> lastTimestampInNanoseconds =
+      new ThreadLocal<Long>() {
+        @Override
+        protected Long initialValue() {
+          return System.nanoTime();
+        }
+      };
+
+  public JavaRuntimeCircuitBreaker(@NonNull final JavaRuntimeCircuitBreakerConfig config) {
+    super(config.getCircuitCheckIntervalInMillis());
+    this.config = config;
+  }
+
+  @Override
+  public boolean isClosed() {
+    if (!isPercentageValueValid(config.getMemoryThreshold())
+        || !isPercentageValueValid(config.getGcCpuThreshold())) {
+      log.warn("Invalid memory threshold configured {}", config.getMemoryThreshold());
+      return false;
+    }
+
+    long currentTimeInNanoseconds = System.nanoTime();
+    long gcCpuTime = getGCCpuTimeNS() - lastTotalGCTimeNS.get();
+    long elapsedTime = currentTimeInNanoseconds - lastTimestampInNanoseconds.get();
+    double gcCpuTimePercentage = (gcCpuTime / (double) elapsedTime) * 100;
+    if (elapsedTime <= 0) {
+      lastTimestampInNanoseconds.set(currentTimeInNanoseconds);
+      lastTotalGCTimeNS.set(gcCpuTime);
+      return false;
+    }
+    double percentageFreeMemory =
+        100 * ((freeMemory() + (maxMemory() - totalMemory())) / (double) maxMemory());
+
+    lastTimestampInNanoseconds.set(currentTimeInNanoseconds);
+    lastTotalGCTimeNS.set(lastTotalGCTimeNS.get() + gcCpuTime);
+
+    int freeMemoryThreshold = config.getMemoryThreshold();
+    int gcCPUThreshold = config.getGcCpuThreshold();
+    log.debug(
+        "Circuit breaker: percentage free memory {}%  GC CPU time percentage {}% (freeMemoryThreshold {}, gcCPUThreshold {})",
+        percentageFreeMemory, gcCpuTimePercentage, freeMemoryThreshold, gcCPUThreshold);
+
+    if (gcCpuTimePercentage >= gcCPUThreshold && percentageFreeMemory <= freeMemoryThreshold) {
+      log.warn(
+          "Circuit breaker tripped at memory {}%  GC CPU time {}%",
+          percentageFreeMemory, gcCpuTimePercentage);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @return CPU time of old gen GC collection in nanoseconds
+   */
+  private long getGCCpuTimeNS() {
+    return TimeUnit.NANOSECONDS.convert(
+        getOldGenGCBean().getCollectionTime(), TimeUnit.MILLISECONDS);
+  }
+
+  private long getGCCount() {
+    long gcCpuCount = 0;
+    long collectorCount;
+    for (GarbageCollectorMXBean gcBean : ManagementFactory.getGarbageCollectorMXBeans()) {
+      collectorCount = gcBean.getCollectionCount();
+      if (collectorCount != -1) {
+        gcCpuCount += gcBean.getCollectionCount();
+      }
+    }
+    return gcCpuCount;
+  }
+
+  /**
+   * Bean with the least amount of gc counts == old gen GC Bean.
+   *
+   * @return GarbageCollector bean for the old (tenured) generation pool
+   */
+  private GarbageCollectorMXBean getOldGenGCBean() {
```

Comment:
Would be nice to document what happens here more, as the GC behavior is somehow cryptic for general Java developer